textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java

```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
     }
 
     /**
+     * When called returns a single-member list containing most-likely tokenization of phrase
+     * Splits phrase into terms by whitespace delimiter
+     * For each term, rewrites(tokenizes) term to detect likely missing spaces
+     * Concatenates all rewritten terms with singlespace delimiter
+     *
+     * For example, if input phrase is "newyork city", it would first split it into "newyork" and "city"
+     * rewriteTerm on "newyork" should return "new york"
+     * rewriteTerm on "city" should return "city"
+     * The two rewritten strings are then concatenated in order by a singlespace to give "new york city"
+     * The result returned as a single-member list is {"new york city"}
+     *
+     * @return
+     */
+    public List<String> getFuzzyTokens() {
+        String[] terms = phrase.split("\\s+");
+        String rewrittenPhrase = "";
+        for(String term : terms)
+            rewrittenPhrase += " " + rewriteTerm(term);
+        rewrittenPhrase = rewrittenPhrase.substring(1);
+        return Arrays.asList(rewrittenPhrase);
+    }
+
+    /**
+     * Performs most likely tokenization of a single term
+     * Uses Dynamic Programming algorithm approach used for tokenization of Chinese strings
+     * Source - Chinese Tokenizer in SRCH2 codebase
+     *           https://github.com/SRCH2/srch2-ngn/blob/master/src/core/analyzer/ChineseTokenizer.cpp#L197
+     * @param term
+     * @return
+     */
+    public String rewriteTerm(String term) {
+        int size = term.length();
+        if(size == 0)
+            return "";
+
+        double[] scoreAtGap = new double[size];
+        int[] preBestGap = new int[size];
+
+        Arrays.fill(scoreAtGap, RewriterConstants.MAXIMUM_SEQUENCE_SCORE);
+        Arrays.fill(preBestGap, -1);
+        scoreAtGap[0] = 0;
+        preBestGap[0] = 0;
+
+        for(int endPosition = 1; endPosition < size; endPosition++) {
+            for(int spanSize = 1; spanSize < RewriterConstants.MAX_WORD_LENGTH && spanSize <= endPosition; spanSize++) {
+                double freq = wordBase.getFrequency(term.substring(endPosition-spanSize, endPosition+1));
+                if(freq < 0) {
+                    if(spanSize == 1)
+                        freq = RewriterConstants.UNKNOWN_CHAR_FREQ;
+                    else
+                        continue;
+                }
+
+                if(freq + scoreAtGap[endPosition-spanSize] + RewriterConstants.TOKEN_LENGTH_PENALTY < scoreAtGap[endPosition]) {
+                    scoreAtGap[endPosition] = freq + scoreAtGap[endPosition-spanSize] + RewriterConstants.TOKEN_LENGTH_PENALTY;
+                    preBestGap[endPosition] = endPosition - spanSize;
+                }
+            }
+        }
+
+        String tokenization = "";
+
+        for(int pos = size-1; pos > 0; pos = preBestGap[pos]-1) {
+            tokenization = " " + term.substring(preBestGap[pos], pos+1) + tokenization;
```

Comment:
In this phase, we reverse-engineer the tokenization by jumping over the preBestGap indices.  
Changing `pos > 0` to `pos > 1` might not help since `1` might not be the index of a likely whitespace. `0` definitely works since the word starts from that index.